graphql.go

```diff
@@ -166,8 +177,29 @@ func (s *Schema) ValidateWithVariables(queryString string, variables map[string]
 	if qErr != nil {
 		return []*errors.QueryError{qErr}
 	}
+	return s.validate(context.TODO(), queryString, doc, variables)
+}
 
-	return validation.Validate(s.schema, doc, variables, s.maxDepth)
+func (s *Schema) validate(ctx context.Context, queryString string, doc *query.Document, variables map[string]interface{}) (errsReturn []*errors.QueryError) {
+	var stats trace.ValidationStats
+	_, finish := s.validationTracer.TraceValidation(ctx)
+	defer func() {
+		finish(errsReturn, stats)
+	}()
+
+	var cacheKey uint64
+	if s.validationCache != nil {
+		cacheKey = s.validationCache.MakeKey(queryString)
```

Comment:
Do we have any way of verifying that the query strings are true constants and that we're not writing any dynamic values into them instead of passing those as arguments? Or do the libraries we use make this impossible?

Reply:
There's no way to *ensure* the query strings are truly constant, either via our libraries or the GraphQL spec itself; GraphQL is designed to allow arbitrary queries.  Also note that this cache relies on the literal text of the query; e.g. adding an extra space or newline, or re-ordering the fields of the query, which have no effect on the semantics of the query, will render the cache useless and result in strictly *more* work.

That said, I've added the `ValidationStats` and plumbed them through via Prometheus metrics in our usage, exactly so we can *verify* what's happening in practice, by monitoring the size of the cache and our hit rate.

My local tests indicate we are indeed sending ~20ish unique query strings via webui+ssrd.  Each different version of {webui+ssrd, extension, mobile app} all might send slightly different query strings.  So I'm expecting the total set of unique queries to be small, like <100.

In concept an "attacker" could also render this cache useless, by purposely sending slightly different query strings.  We can fix that if it becomes a problem.